pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,420 @@
+}
+
+type RawConfig struct {
+	APIKey       providerconfig.ConfigVarString   `json:"apiKey"`
+	ProjectID    providerconfig.ConfigVarString   `json:"projectID"`
+	InstanceType providerconfig.ConfigVarString   `json:"instanceType"`
+	Facilities   []providerconfig.ConfigVarString `json:"facilities"`
+	OS           providerconfig.ConfigVarString   `json:"os"`
+	Tags         []providerconfig.ConfigVarString `json:"tags"`
+}
+
+type Config struct {
+	APIKey       string
+	ProjectID    string
+	InstanceType string
+	Facilities   []string
+	OS           string
+	Tags         []string
+}
+
+type provider struct {
+	configVarResolver *providerconfig.ConfigVarResolver
+}
+
+func (p *provider) getConfig(s v1alpha1.ProviderSpec) (*Config, *providerconfig.Config, error) {
+	if s.Value == nil {
+		return nil, nil, fmt.Errorf("machine.spec.providerconfig.value is nil")
+	}
+	pconfig := providerconfig.Config{}
+	err := json.Unmarshal(s.Value.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rawConfig := RawConfig{}
+	if err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig); err != nil {
+		return nil, nil, err
+	}
+
+	c := Config{}
+	c.APIKey, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.APIKey, "PACKET_API_KEY")
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"apiKey\" field, error = %v", err)
+	}
+	c.InstanceType, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.InstanceType)
+	if err != nil {
+		return nil, nil, err
+	}
+	for i, tag := range rawConfig.Tags {
+		tagValue, err := p.configVarResolver.GetConfigVarStringValue(tag)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get the %d value of \"tags\" field, error = %v", i, err)
+		}
+		c.Tags = append(c.Tags, tagValue)
+	}
+	for i, facility := range rawConfig.Facilities {
+		facilityValue, err := p.configVarResolver.GetConfigVarStringValue(facility)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get the %d value of \"facilities\" field, error = %v", i, err)
+		}
+		c.Facilities = append(c.Facilities, facilityValue)
+	}
+	return &c, &pconfig, err
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	c, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.APIKey == "" {
+		return errors.New("apiKey is missing")
+	}
+
+	_, err = getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	client := getClient(c.APIKey)
+
+	if len(c.Facilities) == 0 || c.Facilities[0] == "" {
+		return fmt.Errorf("must have at least one non-blank facility")
+	}
+
+	// get all valid facilities
+	facilities, _, err := client.Facilities.List(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list facilities: %v", err)
+	}
+	// ensure our requested facilities are in those facilities
```

Comment:
Is the list an expensive API call? Could we instead just GET the ones we are interested in?

Reply:
Actually, there is no option to GET a facility. The only option is to list them. It has them in a static table, so it isn't as if it is collecting them from lots of places. It is (or should be, as I understand it), a simple `N(1)` call.